cputest/fma.cpp

```diff
@@ -0,0 +1,1591 @@
+  if (IsInfinity(b_unpacked) && infinity_sign != b_unpacked.sign)
+    return QNAN;
+  else
+    return infinity;
+}
+
+static u64 TruncateSingleFraction(u64 value, bool single = true)
+{
+  if (single)
+    value &= 0xffff'ffff'e000'0000;
+  return value;
+}
+
+static u32 GetRN(u32 fpscr_low_bits)
+{
+  return fpscr_low_bits & 0x3;
+}
+
+static bool GetNI(u32 fpscr_low_bits)
+{
+  return (fpscr_low_bits & 0x4) != 0;
+}
+
+static u64 SoftwareFma(u64 a, u64 b, u64 c, u32 fpscr_low_bits, bool single)
+{
+  UnpackedFloat a_unpacked = UnpackFloat(a);
+  UnpackedFloat b_unpacked = UnpackFloat(b);
+  UnpackedFloat c_unpacked = UnpackFloat(c);
+
+  if (s_debug_prints)
+  {
+    network_printf(fmt::format("a: {:01x} {:03x} {:016x}\n", a_unpacked.sign, a_unpacked.exponent, a_unpacked.mantissa).c_str());
+    network_printf(fmt::format("b: {:01x} {:03x} {:016x}\n", b_unpacked.sign, b_unpacked.exponent, b_unpacked.mantissa).c_str());
+    network_printf(fmt::format("c: {:01x} {:03x} {:016x}\n", c_unpacked.sign, c_unpacked.exponent, c_unpacked.mantissa).c_str());
+  }
+
+  if (IsNaN(a_unpacked))
+    return TruncateSingleFraction(a | DOUBLE_QBIT, single);
+  else if (IsNaN(b_unpacked))
+    return TruncateSingleFraction(b | DOUBLE_QBIT, single);
+  else if (IsNaN(c_unpacked))
+    return TruncateSingleFraction(c | DOUBLE_QBIT, single);
+
+  const u64 infinity_multiplication_result = InfinityWithSign(a_unpacked.sign ^ c_unpacked.sign);
+  if (IsInfinity(a_unpacked) && IsInfinity(c_unpacked))
+  {
+    return TruncateSingleFraction(AddInfinityAndB(infinity_multiplication_result, b_unpacked), single);
+  }
+  else if (IsInfinity(a_unpacked))
+  {
+    if (c_unpacked.exponent == 0 && c_unpacked.mantissa == 0)
+      return QNAN;
+    else
+      return TruncateSingleFraction(AddInfinityAndB(infinity_multiplication_result, b_unpacked), single);
+  }
+  else if (IsInfinity(c_unpacked))
+  {
+    if (a_unpacked.exponent == 0 && a_unpacked.mantissa == 0)
+      return QNAN;
+    else
+      return TruncateSingleFraction(AddInfinityAndB(infinity_multiplication_result, b_unpacked), single);
+  }
+  else if (IsInfinity(b_unpacked))
+  {
+    return TruncateSingleFraction(InfinityWithSign(b_unpacked.sign), single);
+  }
+
+  // Round the C input for single-precision multiplication
+  if (single)
+  {
+    if (c_unpacked.mantissa != 0)
+    {
+      const u64 round_bit = 0x08000000;
+      if ((c_unpacked.mantissa & round_bit) != 0)
+        c_unpacked.mantissa += round_bit;
+      c_unpacked.mantissa &= ~(round_bit - 1);
+
+      if ((c_unpacked.mantissa & (u64(1) << 52)) != 0)
+      {
+        c_unpacked.mantissa = 0;
+        c_unpacked.exponent += 1;
+      }
+    }
+
+    if (s_debug_prints)
+      network_printf(fmt::format("{:01x} {:03x} {:016x}\n", c_unpacked.sign, c_unpacked.exponent, c_unpacked.mantissa).c_str());
+  }
+
+  a_unpacked = Normalize(a_unpacked);
+  b_unpacked = Normalize(b_unpacked);
+  c_unpacked = Normalize(c_unpacked);
```

Comment:
The normalization of C should happen before this I believe. This is the only reason that for example `0x0000000000000001` doesn't get turned into 0

Reply:
Ah, yet another behavior I didn't know of that Dolphin doesn't emulate correctly 😅